cmds/cpu/cpu9p_unix_test.go

```diff
@@ -0,0 +1,199 @@
+
+	c := &cpu9p{
+		path: f,
+	}
+
+	q, gm, ga, err := c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	t.Logf("First getattr returns (%v, %v, %v)", q, gm, ga)
+
+	sam := p9.SetAttrMask{
+		Permissions:        false,
+		UID:                false,
+		GID:                false,
+		Size:               false,
+		ATime:              false,
+		MTime:              false,
+		CTime:              false,
+		ATimeNotSystemTime: false,
+		MTimeNotSystemTime: false,
+	}
+	sa := p9.SetAttr{
+		Permissions:      0777,
+		UID:              ga.UID,
+		GID:              ga.GID,
+		Size:             ga.Size + 4,
+		ATimeSeconds:     ga.ATimeSeconds + 22,
+		ATimeNanoSeconds: ga.ATimeNanoSeconds + 33,
+		MTimeSeconds:     ga.MTimeSeconds + 55,
+		MTimeNanoSeconds: ga.MTimeNanoSeconds + 66,
+	}
+
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with no change: %v != nil", err)
+	}
+	q, gm, ga2, err := c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	if !reflect.DeepEqual(ga, ga2) {
+		t.Fatalf("Second getattr after empty setattr differs: %v != %v", ga2, ga)
+	}
+
+	sam.Permissions = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with mode: %v != nil", err)
+	}
+	q, gm, ga2, err = c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	if ga2.Mode&p9.AllPermissions != sa.Permissions {
+		t.Errorf("Permissions: %o != %o", ga2.Mode&p9.AllPermissions, sa.Permissions)
+	}
+
+	sam.Size = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with size: %v != nil", err)
+	}
+	q, gm, ga2, err = c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	if ga2.Size != sa.Size {
+		t.Errorf("Size: %o != %o", ga2.Size, sa.Size)
+	}
+
+	sam.UID = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with UID: %v != nil", err)
+	}
+	q, gm, ga2, err = c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	if ga2.UID != sa.UID {
+		t.Errorf("UID: %o != %o", ga2.UID, sa.UID)
+	}
+
+	sam.GID = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with GID: %v != nil", err)
+	}
+	q, gm, ga2, err = c.GetAttr(m)
+	if err != nil {
+		t.Fatalf("Getattr: %v != nil", err)
+	}
+	if ga2.GID != sa.GID {
+		t.Errorf("GID: %o != %o", ga2.GID, sa.GID)
+	}
+
+	// Setting either of these should be ok,
+	// and will set both to current.
+	// Checking is hazardous, as file systems
+	// don't always do what you might expect
+	// around time. We settle for no error.
+	sam.ATime = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with ATime: %v != nil", err)
+	}
+	sam.MTime = true
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with MTime: %v != nil", err)
+	}
+	sam.ATimeNotSystemTime, sam.MTime = true, false
+	if err := c.SetAttr(sam, sa); err != nil {
+		t.Fatalf("Setattr with ATime, sam.ATimeNotSystemTime: %v != nil", err)
+	}
+
+	q, gm, ga2, err = c.GetAttr(m)
```

Comment:
this makes the test stateful and confusing - a simple table test would be clearer with one step at a time

Reply:
(or at least split into multiple parts, as some cases are deliberately making subsequent changes, being inherently stateful)

Reply:
I thought of a table but it's a bit messier than it seems at first.

Reply:
The stateful changes reflect the stateful file change. The problem is the way the mask struct is layed out with its many bool fields, as opposed to a bit mask you could easily modify in for loop. You're either going to get a gigantic test full of struct inits, as you progressively enable more and more mask bits; or you're going to end up using reflect to select fields one by one to set them from false to true. I've laid out code for all cases and believe it or not this is the least wordy version. I'd argue that it's also the least clever, as opposed to using reflect. For now, I don't see the need to make this change.